### ISSUE TITLE
Windows Support

### ISSUE THREAD

**Issue by enko** [dino/dino#309]
This ticket should serve as a kind of meta ticket, to track the efforts needed to bring dino to windows. I have automated the build in #304, so we can put out builds without any problems.

From my small assessment there are two high priority that needs a fix before any other work for the windows port can be done:

* `is_24h_format` from `helper.vala` crashes dino. From my limited knowledge of GTK, this seems to be ok to crash since there is no gnome-functionality on windows?
* Notifications are broken [upstream](https://bugzilla.gnome.org/show_bug.cgi?id=776583). As a possible fix, just disable the notifications on windows, till a proper backend lands in GTK?

On the chat there was a discussion if dino should look native, I think it should, but the first priority should be that it does not crash or behaves funny. Then there should be put energy into making dino look native.

**Comment by Echolon** [comment]
I think it should be native, too. Is this only a question of GUI? Or are there other related issues if we decide for one or another? I mean, if people don't like it - fork it, do it the other way.... or?

**Comment by Echolon** [comment]
Vote native

**Comment by Echolon** [comment]
Vote against native

**Comment by DeSaurus** [comment]
Someone made a dummy win32 notification backend for GNotification
https://bugzilla.gnome.org/show_bug.cgi?id=776583#c7
I hope this gets fixed soon

**Comment by LuccoJ** [comment]
I think a "vote" for/against "native" when it hasn't been determined what "native" means just before the vote is not useful.

**Comment by Echolon** [comment]
@LuccoJ HI, guess you are probably right. Thought it's clear - may someone help out with a good definition?

Anyway, the vote so far made clear that this is polarizing topic. That's an evaluation, too.

**Comment by Echolon** [comment]
@enko Hi, any updates so far? If there are issues, maybe someone else could help?

**Comment by LAGonauta** [comment]
Just FYI, Dino works on WSL with a vcXsrv after intalling all dependencies.
At first it crashed on load, but was fixed after running "sudo dpkg-reconfigure tzdata".

**Comment by Echolon** [comment]
@LAGonauta   Thanks for your info. Any extra info we might need to know? If you can exactly document how you make it run, that would be great!

**Comment by LAGonauta** [comment]
Sure:
**After fresh install of WSL Ubuntu 18.04**

## Update installation:
```
sudo apt update && sudo apt upgrade
```

## Install dependencies:
```
sudo apt install build-essential ninja-build cmake valac libgee-0.8-dev libsqlite3-dev libgdk-pixbuf2.0-dev libgtk-3-dev libsoup2.4-dev libgpgme-dev libqrencode-dev libgcrypt20-dev dbus-x11
```

## Clone and build:
```
sudo apt install git
git clone https://github.com/dino/dino/
cd dino && ./configure
make
```
## Configure timezone data:
```
sudo dpkg-reconfigure tzdata
```
## Install X server:
https://sourceforge.net/projects/vcxsrv/
After installation, run Xlaunch on multi-window mode.
## Run:
Go back to WSL and run dino:
```
export DISPLAY=:0
export LIBGL_ALWAYS_INDIRECT=1
build/dino
```

## Optional dependency:
```
sudo apt install fonts-emojione
```

It is not perfect as for some reason the window does not behave as a normal window on Windows (probably because of the decorations), however all the rest seems to be working fine. Definitely prettier than Gajim :P

**Comment by YtvwlD** [comment]
Do you really need to build dino yourself in this case or could you just add the repository?

**Comment by Echolon** [comment]
@LAGonauta  I guess that NOT for Microsoft Windows, right?

This is what the issue is about

**Comment by LAGonauta** [comment]
Yes, it is Microsoft Windows:
https://docs.microsoft.com/en-us/windows/wsl/install-win10

**Comment by Echolon** [comment]
Ahh, Win7 was my last MS system years ago, never used Win10 for all reasons ;)

@enko experienced big isses with the MS visual studio compling an .exe. Do you (or anyone else) know people with more experience with this? @LAGonauta

However, thanks for your post!

**Comment by LAGonauta** [comment]
![dino!](https://user-images.githubusercontent.com/14236067/74082586-48e8db00-4a3a-11ea-8647-daf05c95e50c.PNG)

With some help from cacahuete from Dino MUC I was able to compile with MSYS2, as shown in the image above 😀
I disabled CSD as it is strange on Windows and not compatible with AeroSnap, but it is working.
Current patch: (for some reason I had to instantiate those classes when compiling on MSYS2)
[instance.txt](https://github.com/dino/dino/files/4173979/instance.txt)

Notifications are not crashing thanks to the changes in glib 👍

Current build script, based #304:
```
#!/bin/bash

cd dino
./configure --with-libsignal-in-tree
make -j5
cd build
mkdir -p dist
cp *.exe *.dll dist
cd dist
ldd dino.exe | grep mingw64 | awk '{print "cp /mingw64/bin/"$1" ."}' | sh

mkdir -p ./share/icons && cp -r /mingw64/share/icons ./share/
mkdir -p ./share/themes && cp -r /mingw64/share/themes ./share/
mkdir -p ./share/glib-2.0/schemas && cp -r /mingw64/share/glib-2.0/schemas ./share/glib-2.0/
mkdir -p ./lib/gdk-pixbuf-2.0/ && cp -r /mingw64/lib/gdk-pixbuf-2.0 ./lib/
mkdir -p ./lib/gio/ && cp -r /mingw64/lib/gio ./lib/
```

It seems that Vala or GTK3 loads most modules dynamically, and `ldd` is not able to find them, thus we must copy them manually. There are still some libraries to be copied, as OMEMO and OpenPGP are not working. Image preview, and opening links/attachments are also not working.

**Comment by licaon-kter** [comment]
Since MAM might be hiding some stuff, I'll post my findings here too, with @LAGonauta build...

It can't find the plugins...
![dinowinplugins](https://user-images.githubusercontent.com/2282799/74094820-720f7700-4adf-11ea-8e89-5a94f4a0b58b.jpg)
...so I create that and put those 3 files there and OMEMO and http_upload are working

Now for PGP...
![dinopgperror](https://user-images.githubusercontent.com/2282799/74094822-789dee80-4adf-11ea-89e6-359d0623f70b.jpg)

...and I have the one from www.gpg4win.org installed but it's in `C:\Program Files (x86)\GnuPG\bin\` I can move stuff around... put it in `C:\Program Files (x86)\GNU\GnuPG\` yet
![dingpgme](https://user-images.githubusercontent.com/2282799/74094824-7b004880-4adf-11ea-9c20-dbd849c43904.jpg)
...ok copy that in the Dino folder... aaaaannnd.... crash...

**Comment by LAGonauta** [comment]
Branch with my tweaks: https://github.com/LAGonauta/dino/tree/master-windows-changes
Newest install dependencies script:
```
#!/bin/bash
pacman -S --needed \
       mingw64/mingw-w64-x86_64-gcc \
       mingw64/mingw-w64-x86_64-cmake \
       mingw64/mingw-w64-x86_64-ninja \
       mingw64/mingw-w64-x86_64-libsoup \
       mingw64/mingw-w64-x86_64-gpgme \
       mingw64/mingw-w64-x86_64-gtk3 \
       mingw64/mingw-w64-x86_64-sqlite3 \
       mingw64/mingw-w64-x86_64-gobject-introspection \
       mingw64/mingw-w64-x86_64-glib2 \
       mingw64/mingw-w64-x86_64-glib-networking \
       mingw64/mingw-w64-x86_64-libgcrypt \
       mingw64/mingw-w64-x86_64-libgee \
       mingw64/mingw-w64-x86_64-pkg-config \
       mingw64/mingw-w64-x86_64-vala \
       mingw64/mingw-w64-x86_64-gsettings-desktop-schemas \
       mingw64/mingw-w64-x86_64-qrencode
```

Newest build script:
```
#!/bin/bash

cd dino && \
    ./configure --with-libsignal-in-tree --no-debug --release && \
    make -j5 && \
    cd build && \
    mkdir -p dist && \
    strip -s dino.exe *.dll && \
    cp *.exe *.dll dist && \
    cd dist && \
    ldd dino.exe | grep mingw64 | awk '{print "cp /mingw64/bin/"$1" ."}' | sh && \
    mkdir -p ./share/icons && cp -r /mingw64/share/icons ./share/ && \
    mkdir -p ./share/themes && cp -r /mingw64/share/themes ./share/ && \
    mkdir -p ./share/glib-2.0/schemas && cp -r /mingw64/share/glib-2.0/schemas ./share/glib-2.0/ && \
    mkdir -p ./lib/gdk-pixbuf-2.0/ && cp -r /mingw64/lib/gdk-pixbuf-2.0 ./lib/ && \
    mkdir -p ./lib/gio/ && cp -r /mingw64/lib/gio ./lib/ && \
    mkdir -p ./plugins/ && mv -f http-files.dll openpgp.dll omemo.dll ./plugins/ && \
    find . -iname "*.dll" -exec ldd {} + | grep mingw64 | awk '{print "cp /mingw64/bin/"$1" ."}' | sh

```

**Comment by gsantner** [comment]
Any chance this gets merged to upstream dino repo, so there will be zip or installer exe available for Dino on Windows?

The biggest amount of friends & family uses Windows and there is no really viable xmpp client at the moment for windows.

**Comment by stevenroose** [comment]
@LAGonauta How do you get the `mingw64/...` packages? It seems that I have to add some database that I don't have.

**Comment by LAGonauta** [comment]
Those commands are exclusive to MSYS2:
https://www.msys2.org/

**Comment by LAGonauta** [comment]
New install dependencies script, adding `ntldd` as `ldd` did not always return all dependencies:
```
#!/bin/bash
pacman -S --needed \
       mingw64/mingw-w64-x86_64-gcc \
       mingw64/mingw-w64-x86_64-cmake \
       mingw64/mingw-w64-x86_64-ninja \
       mingw64/mingw-w64-x86_64-libsoup \
       mingw64/mingw-w64-x86_64-gtk3 \
       mingw64/mingw-w64-x86_64-sqlite3 \
       mingw64/mingw-w64-x86_64-gobject-introspection \
       mingw64/mingw-w64-x86_64-glib2 \
       mingw64/mingw-w64-x86_64-glib-networking \
       mingw64/mingw-w64-x86_64-libgcrypt \
       mingw64/mingw-w64-x86_64-libgee \
       mingw64/mingw-w64-x86_64-pkg-config \
       mingw64/mingw-w64-x86_64-vala \
       mingw64/mingw-w64-x86_64-gsettings-desktop-schemas \
       mingw64/mingw-w64-x86_64-qrencode \
       mingw64/mingw-w64-x86_64-ntldd-git

# GPGME 1.13 is not working (crashes), but 1.12 does. However 1.12 does not have a *.pc file. Let's get it from the 1.13 archive.
mkdir temp-gpgme && \
      cd temp-gpgme && \
      wget http://repo.msys2.org/mingw/x86_64/mingw-w64-x86_64-gpgme-1.12.0-1-any.pkg.tar.xz && \
      pacman -U ./mingw-w64-x86_64-gpgme-1.12.0-1-any.pkg.tar.xz && \
      wget http://repo.msys2.org/mingw/x86_64/mingw-w64-x86_64-gpgme-1.13.1-4-any.pkg.tar.xz && \
      unxz mingw-w64-x86_64-gpgme-1.13.1-4-any.pkg.tar.xz && \
      tar --strip-components=3 -xf mingw-w64-x86_64-gpgme-1.13.1-4-any.pkg.tar mingw64/lib/pkgconfig/gpgme.pc  mingw64/lib/pkgconfig/gpgme-glib.pc && \
      sed -i 's/1.13.1-unknown/1.12.0-unknown/g' *.pc && \
      mv *.pc /mingw64/lib/pkgconfig/ && \
      cd .. && \
      rm -r temp-gpgme
```

New build script that uses `make install` to be easier to maintain:
```
cd dino && \
    ./configure --with-libsignal-in-tree --program-prefix=${PWD}/dist --no-debug --release && \
    make -j5 && \
    make install && \
    cd dist && \
    cp /mingw64/bin/gdbus.exe ./bin && \
    cp /mingw64/bin/gspawn-win64-helper.exe ./bin && \
    mkdir -p ./lib/gdk-pixbuf-2.0/ && cp -r /mingw64/lib/gdk-pixbuf-2.0 ./lib/ && \
    mkdir -p ./lib/gio/ && cp -r /mingw64/lib/gio ./lib/ && \
    mkdir -p ./share/icons && cp -r /mingw64/share/icons ./share/ && \
    mkdir -p ./share/locale && cp -r /mingw64/share/locale ./share/ && \
    mkdir -p ./share/glib-2.0/schemas && cp -r /mingw64/share/glib-2.0/schemas ./share/glib-2.0/ && \
    rm -r ./include && 
    find . -iname "*.dll.a" -exec rm {} + && \
    find . -iname "*.exe" -exec ntldd {} + | grep mingw64 | awk '{print "cp /mingw64/bin/"$1" ."}' | sh && \
    find . -iname "*.dll" -exec ntldd {} + | grep mingw64 | awk '{print "cp /mingw64/bin/"$1" ."}' | sh && \
    find . -iname "*.exe" -exec ldd {} + | grep mingw64 | awk '{print "cp /mingw64/bin/"$1" ."}' | sh && \
    find . -iname "*.dll" -exec ldd {} + | grep mingw64 | awk '{print "cp /mingw64/bin/"$1" ."}' | sh && \
    find . -iname "*.exe" -exec strip -s {} + && \
    find . -iname "*.dll" -exec strip -s {} + && \
    mv *.dll ./bin
```

**Comment by sonicnkt** [comment]
Anyway you could release a precompiled built for testing? 

Im currently in the process of upgrading our xmpp system in our organisation (move to prosody and new client). We have a mixed environment with windows linux and mac clients but i would love to deploy a single desktop client to all users.  Dino seems to be a great solution as it has all features i would like to use built in, the only problem atm is the missing windows support. 

Any timeline on when this will be merged and prebuilt packages available? 0.2 Milestone? 

Currently we are using pidgin but since more users want to be able to use more than one device with all messages synced this wont be a solution for much longer.

**Comment by Echolon** [comment]
@sonicnkt better use anything than Pidgin. If the prebuilds are not gonna make it, evtl. have a look at Gajim, Converse.js or UWPX (Alpha)

**Comment by sonicnkt** [comment]
Yeah i set this up a few years ago and it had a clean ui and worked for everything we needed back then.  Today it wont cut it anymore. 

Didnt really like gajim as well i will take a look at UWPX but im not sure this will work on all clients as we are using win10 LTSB/LTSC releases. The Sideload installer is quite huge with 300MB 😱 

I know this is getting a bit off topic but how does converse handle large file transfers? we are using xmpp alot to transfer huge files p2p internally.

**Comment by langrafik** [comment]
Can anyone just build  .exe installer for Dino please?))

**Comment by mdosch** [comment]
@langrafik 
Afaik the Dino devs will release windows builds once it is considered good enough.

Till this is achieved you can have a look [here](https://github.com/LAGonauta/dino) where someone makes Dino able to compile on Windows. If you managed to build it you can easily create an installer by using NSIS (see my [proof-of-concept](https://github.com/LAGonauta/dino/pull/2).

**Comment by Deluvi** [comment]
Hey everyone! For those who just wants to try out Dino on Windows, I have set up a CI based on Github actions so you don't have to: https://github.com/LAGonauta/dino/pull/3. The PR is currently a draft but it is usable (that's my main XMPP client on Windows atm).
The CI outputs a Windows .exe installer for every release I perform. You can get the most fresh builds there: https://github.com/Deluvi/dino/releases (And maybe someday here: https://github.com/LAGonauta/dino/releases and eventually, with my best hope, here https://github.com/dino/dino/releases 😄)
It is not an official release in any shape or form, just something I hacked over the week-end. It's more of a demo of what LAGonauta achieved in his compatibility tweaks PR: a working version of Dino on Windows!

**Comment by licaon-kter** [comment]
@Deluvi why to PR this CI workflow here?

**Comment by Deluvi** [comment]
To be clear, I am doing my PR on top of LAGonauta's PR. I don't think my PR should be merged before his PR is upstreamed on Dino in order to keep his PR scoped: I will probably rebase my CI PR on the upstream repository.

However, Dino will need a Windows CI eventually. The state of package managers on Windows requires, unfortunately, the upstream developers to ship their own Windows build unlike Linux distributions. How this build will be performed is a needed conversation with the maintainers if we wants Dino to be shipped efficiently on Windows.

As an example, Transmission[1] and Inkscape[2] both uses appveyor to automatically builds their own Windows installer. Afterwards, those installers are picked up by the packages managers (such as Chocolatey) to be installed, no builds are performed by the maintainers of the choco packages[3][4].

I am in no way expert on that field, but it seems that, according to my examples, upstream Dino's maintainers will need to provide an official Windows build if we want to support that platform. A CI build would ease the burden of the maintainers to not build themselves manually the installer for each release while accepting contributions of improvement of the Windows build and ensuring the Windows build will not break in a future version.

[1]: [Link of the CI workflow file](https://github.com/transmission/transmission/blob/master/appveyor.yml), [Link of latest release build](https://ci.appveyor.com/project/transmissionbt/transmission/builds/33049239)
[2]: [Link of the CI workflow file](https://gitlab.com/inkscape/inkscape/-/blob/master/.appveyor.yml), [Link of latest release build](https://ci.appveyor.com/project/inkscape/inkscape/builds/35064527)
[3]: [Link to the Transmission's choco script](https://github.com/tunisiano187/Chocolatey-packages/blob/d4067fd90f5a81898de328119ab963608c6733b4/automatic/transmission/tools/chocolateyInstall.ps1#L8)
[4]: [Link to the Inkscape's choco script](https://github.com/chocolatey-community/chocolatey-coreteampackages/blob/ebb18efeec3c14811efe22e1bacea69f5ede4a7f/automatic/inkscape/tools/chocolateyInstall.ps1#L8)

**Comment by wiktor-k** [comment]
@Deluvi From my memory Dino devs explicitly said they want to officially ship Dino through official Windows Store only and that's what's keeping the lags PR from being merged (you may want to double check this with official dino channel).

**Comment by mdosch** [comment]
You don't need to create an installer for windows store?

**Comment by wiktor-k** [comment]
AFAIK In Windows Store you [publish MSIX packages](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-distribute) and that's used by Windows Store to install the app on end-user system. This is not the same as a regular exe installer app.

**Comment by fiaxh** [comment]
> From my memory Dino devs explicitly said they want to officially ship Dino through official Windows Store only and that's what's keeping the lags PR from being merged

No, that's not our position and I doubt we ever said so. Not everyone wants to use the Windows Store and thus it defitively makes sense to also provide Windows builds ourselves via the Dino Website.

That being said, we want to provide a **well working** Dino build, and not some "mostly ok-ish" version. One important example are notifications, which (to my knowledge) are currently not working on Windows.

**Comment by LAGonauta** [comment]
> 
> 
> > From my memory Dino devs explicitly said they want to officially ship Dino through official Windows Store only and that's what's keeping the lags PR from being merged
> 
> No, that's not our position and I doubt we ever said so. Not everyone wants to use the Windows Store and thus it defitively makes sense to also provide Windows builds ourselves via the Dino Website.
> 
> That being said, we want to provide a **well working** Dino build, and not some "mostly ok-ish" version. One important example are notifications, which (to my knowledge) are currently not working on Windows.

Yes, currently the only notification available is the taskbar flashing. Will only be fixed after this issue is closed:
https://gitlab.gnome.org/GNOME/glib/-/issues/1234

I suppose it would be possible to make something up as a temporary Dino plugin that adds native Windows notification (could be something similar to https://github.com/dino/dino/tree/master/plugins/notification-sound), however I do not know if anyone is willing to do that. I have however in my personal backlog adding notification sound through OpenAL, but currently it is at the bottom of it.

**Comment by guysoft** [comment]
It sounds like there is a working build without notifications.
How about just pushing that so there is something to work with. It might make others (including myself) find the time to take a look and fix that spesific problem. 
Otherwise this might hang here for another 3 more years. It seems like its pretty close now. @Deluvi Noted that its usable daily at this point.

**Comment by licaon-kter** [comment]
@guysoft better join the group and test test test `xmpp:dino-win@chat.diebesban.de?join`

**Comment by guysoft** [comment]
In, can't see you

**Comment by licaon-kter** [comment]
@guysoft I'm not testing :) ask the friendly members

**Comment by trymeouteh** [comment]
+1 for Windows version

**Comment by shkarimpour** [comment]
+1

**Comment by BentiGorlich** [comment]
I just found this repo and am very interested in testing Dino. I am a windows user (I don't have a linux on my main desktop). Can I install and use Dino on windows or is it not working yet? 
Windows is not mentioned on the https://github.com/dino/dino/wiki/Distribution-Packages site

**Comment by licaon-kter** [comment]
@BentiGorlich use the unofficial build for now
https://github.com/LAGonauta/dino - releases

And join their group for more info
`xmpp:dino-win@chat.diebesban.de?join`

**Comment by germanicianus** [comment]
@mar-v-in, this issue is now open for 5 years. @LAGonauta provided Dino releases for Windows now for 3 years. Is there any valid reason left not to provide at least an experimental official Dino release for Windows? That would increase the user base tremendously. I think it's not important to have a native look but only to have it working reliably.

WSL is just a crutch to get Dino running on Windows and no average user is able or willing to fiddle with that - most likely these users are not even aware of WSL.

As far as I'm aware, Dino is the only XMPP client providing support for multip-party video conferences.

Just make it as simple as possible for the average user to get started with Dino. For Windows there should be an installer and/or a portable (self-contained) version which just needs to be extracted - latter possible even as PortableApps version - see https://portableapps.com/ .

**Comment by Neustradamus** [comment]
In more, there is a GSoC too:
- https://hrxi.github.io/gsoc/

**Comment by RageGamerBoi** [comment]
I can’t believe this isn’t ready yet.
Everything looks functional, and I couldn’t care less about the “non-native” ui, I just want a functioning Dino build for Windows so I can use it with other people I know.

**Comment by wiktor-k** [comment]
@RageGamerBoi while I can relate to "wouldn't it be nice to have Dino on more OSes" bugging people here is unlikely to change anything. Maybe the constructive way forward would be to try the *unofficial* builds of Dino for Windows:

- https://github.com/LAGonauta/dino/releases
- https://github.com/mxlgv/dino/releases

(FTR I'm not affiliated with any of them and also won't guarantee their quality :man_shrugging:)

**Comment by RageGamerBoi** [comment]
I use Linux myself and I'm not gonna go say "go download this unofficial build of an app you've never head of before" to other people on Windows. That's simply not a good user experience.

**Comment by wiktor-k** [comment]
Then *subscribe* to this ticket and just *wait*. I guess it's quite obvious to everyone it's "not a good user experience" but take my word that "I just want" type of comments do *not* make it happen even a second faster.

**Comment by RageGamerBoi** [comment]
It’s been 3 years and nothing has happened, it’s already done and just needs to be integrated. Let me remind you, it’s been like that for 3 *years*.
When has “subscribe and wait” worked for something like that?
At least somebody is thinking about it again.
(And as far as effect, my comment was no different from the 3 other “+1” comments)

**Comment by Neustradamus** [comment]
A lot of people wait the official support since a long time, and there were forks and Google Summer of Code too.

6 years since the ticket creation.

**Comment by silberzwiebel** [comment]
For more information on the GSOC project 2023, see https://hrxi.github.io/gsoc/blog/gsoc-2023-conclusion

**Comment by aehimself** [comment]
As Skype is going to be officially retired I quickly set up a self-hosted XMPP server. I found Dino which I am extremely happy for and it runs wonderfully on Linux. My family uses Windows though which makes communication a "bit" harder.
The well-known clients on Windows still don't support audio and/or video calls or look extremely outdated and simply inconvenient to use.

The unofficial build from LAGonauta does not start, only half of an empty window shows up.
mxlgv's build works perfectly but it's more out-of-date and some functions aren't working (webcam is not detected, noise cancellation does not work, notifications aren't shown) and unfortunately cannot be built again as some dependencies are out-of-date.

I would also LOVE to see an official Dino release for Windows as it would be the first which supports audio-video calls on this platform.

Keep up the awesome work and thank you for bringing us the best XMPP client! (hopefully, for Windows too, soon :) )

**Comment by Neustradamus** [comment]
I think that we must to contact all people who would like a Dino Windows release to comment and to add a reaction in the main description of this ticket.

For the moment in Windows XMPP Client, there are:
- Psi (not recently updated)
- Psi+ (recently updated)
- Gajim (but not the very recent Gajim 2.0.0 yet -> 1.9.5)

**Comment by mjohnson9** [comment]
@Neustradamus wrote,
> I think that we must to contact all people who would like a Dino Windows release to comment and to add a reaction in the main description of this ticket.

Your threats of brigading are misplaced here. They're unlikely to have anything but negative effects on the development on this feature and the broader project.

You are apparently a contributor to the radvd project[^1][^2], so you presumably should be aware of how open source, community-run projects operate.

As another member of the open source community who is otherwise unaffiliated with this project: **Knock it off and stop threatening to harass people.**

[^1]: https://github.com/orgs/radvd-project/people
[^2]: https://github.com/radvd-project/radvd/commits?author=Neustradamus

**Comment by Neustradamus** [comment]
@mjohnson9: I am in several projects yes :)

I follow the recent comment done by @aehimself who explains the history of a Dino for Windows fork.
A lot of people wait an official Windows version.

There was a GSoC about Windows support (@hrxi):
- https://hrxi.github.io/gsoc/
- https://hrxi.github.io/gsoc/blog/windows-support-for-dino
- https://hrxi.github.io/gsoc/blog/windows-woes
- https://hrxi.github.io/gsoc/blog/gsoc-2023-conclusion

It is important to show that:
- users would like a Windows version
- team needs help and search people to contribute

Note: There were two forks:
- https://github.com/mxlgv/dino/releases
- https://github.com/LAGonauta/dino/releases

**Comment by mar-v-in** [comment]
As you wrote yourself, Windows support is being worked on. For example, the transition to Meson as a build system is part of the strategy to add proper support for Windows. But things will take time.

It really does not help for people to express more deeply that they want Windows support. It's not going to change how we work on this. Dino developers work on Dino in their spare time, which is limited, and while we see the need for Windows support, doing it is far less interesting or fun than other things we do in Dino.

Pushing people with words and public blame to do things in open source is nothing else than telling them to work unpaid overtime. There's only two ways here: Let people do the work in the way and the pace they want - or pay them appropriately to do work the way and pace you want.

If anyone wants to contribute on that matter, that of course would be nice and they're invited to get in touch to see where they can help best - but I bet that was obvious already.

**Comment by aehimself** [comment]
So the good news is that I managed to get mxlgv's version to compile. I'll attempt to roll back any enhancements and rebase it to the current Dino codebase to have a vanilla 0.4.5 Windows version.

Wish me luck :)

**Comment by Neustradamus** [comment]
@marvin: It is notified in my message, that any help is welcome :)

@aehimself: You are welcome, good luck, please inform us!

**Comment by Neustradamus** [comment]
@mar-v-in: It is notified in my message, that any help is welcome :)

@aehimself: You are welcome, good luck, please inform us!

**Comment by aehimself** [comment]
Soooo... after endless merge conflicts I abandoned the idea to rebase mxlgv's changes to the upstream. I went back and started working on LAGonauta's version. As a bonus, there's really recent activity in the past couple of days there!

Version number is not accurate, it is actually compiled from the current Dino repository:
![Image](https://github.com/user-attachments/assets/9ebc14c4-068b-423c-b698-0c5a9ce03a1c)

Took me more time than probably should but this was my first time dealing with MINGW or meson...
I'll clean the dev system up and write a step-by-step how-to on how to compile Dino on Windows, for Windows with MINGW.

**Comment by aehimself** [comment]
To compile Dino on Windows for yourself, download and install MINGW and run the following script:

```
#!/bin/bash

dinoroot=/D/Dino
mingwroot=/C/msys64/mingw64
mergebranch=windowsmerge

#
# From this point you should not need to modify anything...
#

dinogit=$dinoroot/dino
dinobuild=$dinogit/build
dinodist=$dinogit/dist
dinoplug*indist=$dinodist/plugins

pacman -S --needed --noconfirm git mingw-w64-x86_64-meson mingw-w64-x86_64-gcc mingw-w64-x86_64-vala mingw-w64-x86_64-libgee mingw-w64-x86_64-gtk4 make zip unzip curl mingw-w64-x86_64-toolchain mingw-w64-x86_64-cmake mingw-w64-x86_64-ninja mingw-w64-x86_64-libsoup3 mingw-w64-x86_64-sqlite3 mingw-w64-x86_64-gobject-introspection mingw-w64-x86_64-glib2 mingw-w64-x86_64-glib-networking mingw-w64-x86_64-openssl mingw-w64-x86_64-libgcrypt mingw-w64-x86_64-gsettings-desktop-schemas mingw-w64-x86_64-qrencode mingw-w64-x86_64-ntldd-git mingw-w64-x86_64-gpgme mingw-w64-x86_64-libadwaita mingw-w64-x86_64-gspell mingw-w64-x86_64-enchant mingw-w64-x86_64-hunspell mingw-w64-x86_64-iso-codes mingw-w64-x86_64-gst-plugins-base mingw-w64-x86_64-gst-plugins-good mingw-w64-x86_64-gst-plugins-bad mingw-w64-x86_64-cppwinrt mingw-w64-x86_64-webrtc-audio-processing mingw-w64-x86_64-webrtc-audio-processing-1 mingw-w64-x86_64-nsis mingw-w64-x86_64-libsignal-protocol-c mingw-w64-x86_64-fontconfig

if [ ! -d $dinogit ]; then
  cd $dinoroot

  git clone https://github.com/dino/dino.git
fi

cd $dinogit

result=$(git branch | grep "* " | grep -i $mergebranch)

if [ $? != 0 ]; then
  git branch $mergebranch
  git checkout $mergebranch
fi

git pull --rebase https://github.com/LAGonauta/dino.git master-windows-changes

if [ -d $dinobuild ]; then
  rm -rdf $dinobuild
fi

meson setup $dinobuild --reconfigure
meson compile -C $dinobuild

if [ -d $dinodist ]; then
  rm -rdf $dinodist
fi

mkdir $dinodist
cd $dinodist
mkdir $dinodist/bin $dinodist/etc $dinodist/lib $dinodist/share $dinodist/lib/dino $dinodist/bin/plugins
find ../build/plugins -type f -name "*.dll" -exec cp {} $dinodist/bin/plugins \;
find ../build -type f -name "*.dll" -exec cp {} $dinodist/bin/ \;
find ../build -type f -name "dino.exe" -exec cp {} $dinodist/bin/ \;
cp $mingwroot/bin/gdbus.exe $dinodist/bin
cp $mingwroot/bin/gspawn-win64-helper.exe $dinodist/bin
cp -r $mingwroot/share/xml $dinodist/share
cp -r $mingwroot/lib/enchant-2 $dinodist/lib
cp -r $mingwroot/lib/gstreamer-1.0 $dinodist/lib
cp -r $mingwroot/etc/fonts $dinodist/etc
mkdir -p $dinodist/lib/gdk-pixbuf-2.0/
cp -r $mingwroot/lib/gdk-pixbuf-2.0 $dinodist/lib/
mkdir -p $dinodist/lib/gio/
cp -r $mingwroot/lib/gio $dinodist/lib/
mkdir -p $dinodist/share/icons
cp -r $mingwroot/share/icons $dinodist/share/
mkdir -p $dinodist/share/locale
cp -r $mingwroot/share/locale $dinodist/share/
mkdir -p $dinodist/share/glib-2.0/schemas
cp -r $mingwroot/share/glib-2.0/schemas $dinodist/share/glib-2.0/
cp $mingwroot/bin/libcrypto-*-x64.dll $dinodist/bin
cp $dinodist/bin/dino.exe $dinodist/bin/dino_with_console.exe

# IMAGE_SUBSYSTEM_WINDOWS_GUI = 0x0002
# IMAGE_SUBSYSTEM_WINDOWS_CUI = 0x0003
# SUBSYSTEM_OFFSET = 0xDC (220)
printf '\x02\x00' | dd of="$dinodist/bin/dino.exe" bs=1 seek=220 count=2 conv=notrunc

find . -iname "*.dll.a" -exec rm {} +
find . -iname "*.exe" -exec ntldd {} + | grep -i "mingw64" | awk '{print $1}' | while read -r line; do cp -n $mingwroot/bin/$line bin; done
find . -iname "*.dll" -exec ntldd {} + | grep -i "mingw64" | awk '{print $1}' | while read -r line; do cp -n $mingwroot/bin/$line bin; done
find . -iname "*.exe" -exec ldd {} + | grep -i "mingw64" | awk '{print $1}' | while read -r line; do cp -n $mingwroot/bin/$line bin; done
find . -iname "*.dll" -exec ldd {} + | grep -i "mingw64" | awk '{print $1}' | while read -r line; do cp -n $mingwroot/bin/$line bin; done
find . -iname "*.exe" -exec strip -s {} +
find . -iname "*.dll" -exec strip -s {} +
```

It will clone the official Dino repository, create a custom branch and rebases it onto LAGonauta's master-windows-changes branch. At the moment they are up-to-date, but can come in handy in the future. Then it installs all dependencies and compiles Dino with Meson. Once it is done, the necessary files are then put in the "dist" folder where you can run it from directly.
The whole script is based on the original GitHub action, with bits taken from Mlxgv (to patch the exe). I also had to correct the plugins directory location as Dino didn't find them.

I have some issues with portability though, the compiled binary runs on my dev PC but crashes in libvulkan when copied to a different PC. Had no chance to test audio / video calls and file transfers yet.

**Comment by 2disbetter** [comment]
> I have some issues with portability though, the compiled binary runs on my dev PC but crashes in libvulkan when copied to a different PC. Had no chance to test audio / video calls and file transfers yet.

I would be happy to test it for you. I'm not building on Linux as I'm a Window's dev, but I am working on it. If you have a binary for me, I'll happily test it for you.

**Comment by aehimself** [comment]
The script above should be executed on Windows, inside the MINGW environment. It will compile and build Dino for you for Windows, on Windows.

**Comment by 2disbetter** [comment]
So the script appears to work but I had to change this: dinoplug*indist

 to just dinoplugindist. 

I can report that it does then compile and start on my system. I tried a video call and can report that the web cam is not found or allowed to be used still. 

Some improvements however over mxlgv's build is that now the borders of the windows do not extend beyond the chat client itself.

**Comment by aehimself** [comment]
As it turns out Dino will never find it's plugins, because it's looking for them at the wrong place. I made this snipplet with Process Explorer:

![Image](https://github.com/user-attachments/assets/4c328806-c84f-4104-9a28-7f80a93e9e86)

Emphasis on the path "C:\Users\Administrator\AppData\Roaming\Dino\C:\msys64\mingw64\lib\dino\plugins" which is obviously invalid. I looked up the plugins loader and it turns out it is also considering the DINO_PLUGIN_DIR environment variable if it's set. Once I did that, plugins were discovered and loaded successfully:

SET DINO_PLUGIN_DIR=%APPDATA%\Dino\lib\dino\plugins

![Image](https://github.com/user-attachments/assets/c2f6d1f3-2d4c-419c-97f3-9e83667ca02a)

I guess it's a Meson issue, just keep this in mind if you want to run it on Windows.

P.s.: Dino built this way still crashes on Windows 11 in vulkan-1.dll.

**Comment by aehimself** [comment]
Based on master-windows-changes by @LAGonauta I compiled a fresh Dino 0.5.0. You must not trust random binaries on the Internet so I urge everyone to compile it for themselves but it can be downloaded [here](https://www.aecentral.org/dino/dino_0.5.0.exe). [VirusTotal](https://www.virustotal.com/gui/file/105aba2e3c06faf29e219254820bd105e0748e28b0fce43cf5ca72db1df644fb).

**Comment by Neustradamus** [comment]
@aehimself: Nice, good job!
Where is your repo?

It is possible to add all in this upstream?

ALL the World would like official x86 and x64 release buids.

**Comment by aehimself** [comment]
The script does not use any upstream for the $mergebranch branch therefore the merged repository is local only. Considering that not neither Dino nor Windows compatibility is my work I don't want to publish it... I barely managed to cut the important bits together so it compiles :)

Only #851 is needed to make this happen; it's 5 years old tho so I guess it won't be merged anytime soon. *In theory* even GitHub actions are included so everything could be automated. But, in the moment they accept the devs officially agreed on supporting Windows platfom in the codebase which I can see they are not eager to.

Until then I can keep building Windows binaries but they aren't official and out of support anyway (I never touched Vala, MINGW or Meson up until this point so if a problem arises I probably won't be able to fix it).
It's better than nothing I believe.

**Comment by LAGonauta** [comment]
I've made some progress in my test branch [here](https://github.com/LAGonauta/dino/tree/master-windows-changes-test-ci). Thanks to aehimself findings the builds should now be portable to other systems :)
On meson, the font plugin has a resource bug (I am certainly not embedding the resource correctly) so it just doesn't work, and a command prompt is shown when Dino loads.
The cmake build seems to be perfect, though. As soon as I finish fixing meson I should generate a new installer for Windows, unfortunately not 0.5.0 because libomemo-c is not available in MSYS2 yet.

**Comment by mar-v-in** [comment]
> unfortunately not 0.5.0 because libomemo-c is not available in MSYS2 yet.

With Meson, you don't need to rely on MSYS to have the packages available anymore, you can build missing dependencies from wrap files. As libomemo-c already supports Meson, this should be very easy.

**Comment by aehimself** [comment]
I finally got libomemo-c to be visible for Meson under MINGW64 so I rebuilt a new Dino 0.5.0 now with OMEMO support. Link is the same, Virustotal link updated.

See the necessary steps [here](https://github.com/LAGonauta/dino/issues/21).

**Comment by aehimself** [comment]
The crash on Windows 11 can be solved by simply removing vulkan-1.dll from the bin folder. I suppose as it was built on a different version of Windows it's incompatible. However, this is a part of the operating system already.
As it only can cause issues there's no reason to ship it with Dino.

**Comment by aehimself** [comment]
I think @LAGonauta scheduled automated builds, I just don't know if artifacts can be downloaded; see [here](https://github.com/LAGonauta/dino/actions/runs/14489755654). The latest Dino works perfectly on Windows, I only have an issue with webcam detection on some PCs (see [here](https://github.com/LAGonauta/dino/issues/22)). I did some extremely basic logging and realized that the device was not even discovered, it wasn't just filtered out from code. Since discovery is handled by libgstreamer (I think...?) there's nothing much we can do about it :(

**Comment by madmini** [comment]
The artifacts from @LAGonauta's automated builds can be downloaded as follows:
From the repo go to [Actions](https://github.com/LAGonauta/dino/actions/workflows/build.yml), select some build, click one of the `Build Windows (...)` jobs, open the `Upload Windows dino installer` step log and click the `Artifact download URL` at the bottom.

**Comment by aehimself** [comment]
0.5.1 can be built with the same script I posted above. [VirusTotal](https://www.virustotal.com/gui/file/c1c8a2d7e49d37d429108b7ec5114eb348bb3bd1e9ce9c5812dfc805655d67ba) got 2 hits, proably because of the self-extract archive...

Edit: Tested with Conversations. Unfortunately video call still do not work, Dino can not detect any webcameras.
Starting with debug also throws an error:

GLib-CRITICAL (recursed) **:
g_win32_pop_invalid_parameter_handler: assertion
'handler->pushed_handler == popped_handler' failed

**Comment by Neustradamus** [comment]
If several people would like to use on Windows OS, there is DinoX:
- https://github.com/rallep71/dinox

cc: @rallep71.

**Comment by aehimself** [comment]
> If several people would like to use on Windows OS, there is DinoX:

Just keep in mind that it was thrown together with AI. That can raise moral and/or security concerns at some.
